### ISSUE TITLE
[FR] Move attachments and images to a generic files model

### ISSUE THREAD

**Issue by matmair** [inventree/InvenTree#5703]
### Please verify that this feature request has NOT been suggested before.

- [X] I checked and didn't find a similar feature request

### Problem statement

Currently, files are handled by different attachment models and as "dumb" image attachments. This makes reusing and working with these files from code (core, plugins) difficult.

### Suggested solution

Create a first-class citizen model for files with metadata and tags support. Migrate all attachments and images over.

```[tasklist]
### Files/Attachment related issues
- [ ] #5691
- [ ] #5429
- [ ] #5531
- [ ] #5048 
- [ ] #5047
- [ ] https://github.com/inventree/InvenTree/issues/2585
```

```[tasklist]
### Out of scope
- [ ] #5532
```

Might also be interesting for #3101

### Describe alternatives you've considered

Create an api-layer that simulates the same thing.

### Do you want to develop this?

- [X] I want to develop this.

## Upvote & Fund

- We're using [Polar.sh](https://polar.sh/inventree) so you can upvote and help fund this issue.
- We receive the funding once the issue is completed & confirmed by you.
- Thank you in advance for helping prioritize & fund our backlog.

<a href="https://polar.sh/inventree/InvenTree/issues/5703">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/inventree/InvenTree/issues/5703/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/inventree/InvenTree/issues/5703/pledge.svg">
</picture>
</a>

**Comment by wolflu05** [comment]
I fully support this idea. This also enables us to use the Django files api with different backends rather than fiddling directly with the fs.

**Comment by matmair** [comment]
This is marked as `horizon`, I will probably only get this done by December or later.

**Comment by matmair** [comment]
@wolflu05 that is part of the difficulty, there are some patterns around that behaviour that are core to how reports work (currently)

**Comment by matmair** [comment]
Started work on this, targeting [0.15.0](https://github.com/inventree/InvenTree/milestone/62)

**Comment by Petrox** [comment]
I would point out that one file might be uploaded multiple times (eg. same picture or safety manual for 2000 pipes with different dimensions and colors maybe) so "id by content hash" and multiple references to the hash could solve deduplication from day one.  Maybe secondary uploads could have different metadata, but include a pointer, that "this is the same file as id=12312345".

**Comment by matmair** [comment]
@Petrox I do not plan to support that, that issue was intentionally left out of this.

**Comment by Petrox** [comment]
Ok, no problem. I'm not making the decisions, just pointed at the low hanging fruit.

**Comment by matmair** [comment]
I think you underestimate the amount of edge cases and (hopefully) atomic database transactions that need to be handled correctly to implement this safe, reliable and performant.
Anyone is free and encouraged to submit a PR implementing it. I am not against the idea but do not see it being valuable enough for the broader user base to spend a bunch of time on it.

**Comment by Petrox** [comment]
I don't think it would be so much more complex compared to the basic option
of "don't look just store", but I do accept that development is incremental
and smaller steps are better than huge ones.

Also file deduplication could be handled at filesystem level if it's
important.

Or maybe it could be a feature 3 years from now. Definitely not critical.

**Comment by matmair** [comment]
A few points here are still open

**Comment by matmair** [comment]
looking into this again @SchrodingersGat @wolflu05 what is your opinion on renaming `attachments` to `files`?

**Comment by wolflu05** [comment]
Good idea, but then we need to first separate the concept of a link attachment and a file.

**Comment by matmair** [comment]
How so? Linked files seem very helpful in general

**Comment by wolflu05** [comment]
Are these links always used for links to files? I actually never have used them for this case, as I always have downloaded the datasheet and uploaded it to inventree in case it gets removed from the internet, but this makes sense when thinking like this.

But if we rename it again, plugins that use attachments will break again. I think we should start treating such thinks like model names, function names, ... as stable things instead of renaming and moving them over and over.

**Comment by SchrodingersGat** [comment]
@matmair can you help me understand where a more "generic" files model would be implemented? I had thought that the changes made already (bringing all attachments into a single table) was sufficient.

**Comment by matmair** [comment]
This would be a good target for 2.0; handling of files is implemented separately across different models. Unifying attachments was a great steps but there are other files in the system.
Handling of files should also be exposed to plugins (like parsing them ie wireviz, displaying them ie. 3d model view. providing previews)

**Comment by Petrox** [comment]
Should you think about file storage of the future I would also quietly
mention ML, LLM processing of files.

(Which could be implemented by plugins, but they will need to add tags,
json files, processing statuses, results, all linked to the original files
in some way).

So tagging, linking, (colorful) labeling, classification, extraction, all
could be a reasonable need in the not too distant future.

Just my two cents. Feel free to ignore.
